ash/src/extensions/khr/present_wait.rs

```diff
@@ -0,0 +1,47 @@
+use crate::prelude::*;
+use crate::vk;
+use crate::{Device, Instance};
+use std::ffi::CStr;
+use std::mem;
+
+#[derive(Clone)]
+pub struct PresentWait {
+    handle: vk::Device,
+    fns: vk::KhrPresentWaitFn,
+}
+
+impl PresentWait {
+    pub fn new(instance: &Instance, device: &Device) -> Self {
+        let fns = vk::KhrPresentWaitFn::load(|name| unsafe {
+            mem::transmute(instance.get_device_proc_addr(device.handle(), name.as_ptr()))
+        });
+        Self {
+            handle: device.handle(),
+            fns,
+        }
+    }
+
+    #[doc = "<https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkWaitForPresentKHR.html>"]
+    pub unsafe fn wait_for_present(
+        &self,
+        swapchain: vk::SwapchainKHR,
+        present_id: u64,
+        timeout: u64,
+    ) -> VkResult<()> {
+        self.fns
+            .wait_for_present_khr(self.handle, swapchain, present_id, timeout)
+            .result()
```

Comment:
Heh I should probably clean this up in the rest of the code as I've been using `.into()` here for the most part whereas some extension code is using `.result()`.

Reply:
I'll do a cleanup pass in all extensions.

Reply:
Separate PR please.

Reply:
Sure.
The cleanup will include:
- calling `into()` instead of `result()`,
- renaming the function pointer members to `fns`,
- moving `fn name()` down with the `fp()` and all.

Reply:
Not sure about moving `fn name()` down.
The recently added extension `dynamic_rendering` has the `fn name()` and `fn fp()` near the top.
What is the preferred layout ?

Reply:
> * calling `into()` instead of `result()`,

I think calling `.result()` everywhere is more clear (and perhaps removing the `From<vk::Result> for VkResult<()>` entirely). Afaik Rust can't deduce the type when using this with a `?` operator in `.into()?;` cases.

CC @Ralith I presume you prefer `.result()` too instead of having to guess what `From`/`Into` has been implemented?

> * renaming the function pointer members to `fns`,

Lovely, that annoyed me when writing `maintenance4` and `dynamic_rendering`

> * moving `fn name()` down with `fn fp()` and all.

Yes please, my eye also fell on that.

> Not sure about moving `fn name()` down.
> The recently added extension `dynamic_rendering` has the `fn name()` and `fn fp()` near the top.

Yeah that annoyed me when writing `maintenance4`, especially after noticing I put it in the wrong place myself in `dynamic_rendering`. Let's have:

- `fn new()`;
- All the commands;
- `fn name()`;
- `fn fp();
- `fn instance/device()`.

Reply:
Exactly.

Reply:
> I think calling .result() everywhere is more clear (and perhaps removing the From<vk::Result> for VkResult<()> entirely). Afaik Rust can't deduce the type when using this with a ? operator in .into()?; cases.

There are 26 locations, outside of extensions, where `into()` is called (24 in device.rs and 2 in instance.rs).

Reply:
> outside of extensions

Those files are also hand-written helpers, just like the extensions.

Reply:
> Those files are also hand-written helpers, just like the extensions.

Yes, they are handled here : https://github.com/MaikKlein/ash/pull/495.
Should have linked the PRs earlier...